### PR TITLE
Add add_weighted_adapter to IA3 adapters

### PR DESCRIPTION
**Pull request by alexrs**
(Partially) Resolves: https://github.com/huggingface/peft/issues/1688
See https://github.com/huggingface/peft/pull/980 for context.

### What
$(IA)^3$ adapters can't be combined. This option, however, is available for other PEFT adapters such as LoRA.

### Solution
Implement `add_weighted_adapter` method for $(IA)^3$ models supporting only weighted average of adapters for now.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
@alexrs Thanks a lot for reviving the feature and updating it so quickly. LMK when you think it's ready for review.

@Abdullah-kwl If you could give this a try, that would be great.

File: src/peft/tuners/ia3/model.py [hunk 1]
```diff
@@ -279,17 +280,20 @@ def set_adapter(self, adapter_name: str | list[str]) -> None:
...
         self.active_adapter = adapter_name
 
-    def _prepare_adapter_config(self, peft_config, model_config):
+    @staticmethod
+    def _prepare_adapter_config(peft_config, model_config):
         if peft_config.target_modules is None:
             if model_config["model_type"] not in TRANSFORMERS_MODELS_TO_IA3_TARGET_MODULES_MAPPING:
                 raise ValueError("Please specify `target_modules` in `peft_config`")
-            peft_config.target_modules = TRANSFORMERS_MODELS_TO_IA3_TARGET_MODULES_MAPPING[model_config["model_type"]]
+            peft_config.target_modules = set(
```
Comment (alexrs) [inline src/peft/tuners/ia3/model.py:288]: As discussed in https://github.com/huggingface/peft/pull/980#discussion_r1356720557, this seemed to be a bug.

File: src/peft/tuners/lora/model.py [hunk 2]
```diff
@@ -575,9 +575,6 @@ def add_weighted_adapter(
 
         if adapter_name in list(self.peft_config.keys()):
             return
-        for adapter in adapters:
-            if adapter not in list(self.peft_config.keys()):
-                raise ValueError(f"Adapter {adapter} does not exist")
```
Comment (alexrs) [inline src/peft/tuners/lora/model.py:580]: This is done in `_check_add_weighted_adapter ` https://github.com/huggingface/peft/blob/77b7238b907c85f21ec756176351d85faf2aeaad/src/peft/tuners/lora/model.py#L462-L464, which is called right after

File: src/peft/tuners/ia3/model.py [hunk 3]
```diff
@@ -393,3 +397,94 @@ def delete_adapter(self, adapter_name: str) -> None:
...
+        key_list = [key for key, _ in self.model.named_modules() if self.prefix not in key]
+        for key in key_list:
+            _, target, _ = _get_submodules(self.model, key)
+            if isinstance(target, IA3Layer):
+                if adapter_name in target.ia3_l:
+                    target_ia3_l = target.ia3_l[adapter_name]
+                else:
+                    continue
+
+                target_ia3_l.data = target_ia3_l.data * 0.0
```
Comment (alexrs) [inline src/peft/tuners/ia3/model.py:484]: I guess here we could use:

```suggestion
                target_ia3_l.data = target_ia3_l.data.zero_()
```

but I tried to follow the code style used in LoRA: https://github.com/huggingface/peft/blob/77b7238b907c85f21ec756176351d85faf2aeaad/src/peft/tuners/lora/model.py#L612

https://pytorch.org/docs/stable/generated/torch.Tensor.zero_.html#torch-tensor-zero

**Comment by alexrs** [comment]
@BenjaminBossan Ready for review! 👀

File: src/peft/tuners/ia3/model.py [hunk 4]
```diff
@@ -279,17 +280,20 @@ def set_adapter(self, adapter_name: str | list[str]) -> None:
                 module.set_adapter(adapter_name)
         self.active_adapter = adapter_name
 
-    def _prepare_adapter_config(self, peft_config, model_config):
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply src/peft/tuners/ia3/model.py:283]: Thanks for fixing this and the set type below.

File: src/peft/tuners/ia3/model.py [hunk 5]
```diff
@@ -393,3 +397,94 @@ def delete_adapter(self, adapter_name: str) -> None:
...
+        target_module_types = {type(self.peft_config[adapter].target_modules) for adapter in adapters}
+        feedforward_module_types = {type(self.peft_config[adapter].feedforward_modules) for adapter in adapters}
+        if len(target_module_types) > 1 or len(feedforward_module_types) > 1:
+            raise ValueError("All adapter configs should have the same type for target and feedforward modules.")
+
+        # Combine target and feedforward modules
+        if str in target_module_types:
+            new_target_modules = "|".join(f"({self.peft_config[adapter].target_modules})" for adapter in adapters)
+        else:
+            new_target_modules = set.union(*(self.peft_config[adapter].target_modules for adapter in adapters))
```
Comment (BenjaminBossan) [inline src/peft/tuners/ia3/model.py:428]: Oh yeah, simpler than the reduce op we have in LoRA :+1:

File: src/peft/tuners/ia3/model.py (hunk 3 above)
Comment (BenjaminBossan) [reply src/peft/tuners/ia3/model.py:484]: Probably `.zero_()` is more efficient (or has the potential to be so), so I'd be fine with this change.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks a lot for your work. This already looks quite good. I have a few smaller comments, and then 3 more general points:

1. I haven't reviewed the tests yet. I see that you refactored `_test_weighted_combination_of_adapters` to work with both LoRA and IA³. However, all the changes make the review quite hard and there is also the risk to accidentally break the LoRA tests (e.g. skipping to test something that is currently tested). Instead, it would be better IMO if `_test_weighted_combination_of_adapters` would call a submethod `_test_weighted_combination_of_adapters_lora`, `_test_weighted_combination_of_adapters_ia3`, etc., with `_test_weighted_combination_of_adapters` being the current function body. Then you can copy and modify it for IA³. Yes, this will result in more code duplication, but for tests that doesn't matter as much and it will be more maintainable, especially since the different methods will always be quite different and thus tests will differ.
2. IIUC, right now, if I merge, say, 10 adapters, all with `weight=1`, the resulting merged adapter will possibly have very large values, right? So I should probably assign `weight=1/10` so that the norm does not shift. Maybe this should be documented.
3. How about adding a section to https://github.com/huggingface/peft/blob/main/docs/source/developer_guides/model_merging.md?

**Comment by alexrs** [comment]
Hey @BenjaminBossan. I made some changes:
- Re-factored the tests to include `_test_weighted_combination_of_adapters_lora` and `_test_weighted_combination_of_adapters_ia3`. These methods are called from `_test_weighted_combination_of_adapters`.
- Added a short section to the docs. Let me know if I should expand it.

With respect to whether we should normalize the adapter weights or not, I'm not sure what is the best approach. On the one hand, I agree that having a combination of adapters with the sum of weights > 1 might not return the best results. On the other hand, I also think it is interesting for users to specify how they want their adapter weights to be normalized. In your example, `weight=1/10` is a totally valid approach, but also is taking a softmax of the weights or passing some weights as 0 if we want to do some sort of Top K combination. I'm not sure this method should be in charge of handling all of that.

Maybe we should just specify in the docs that we recommend `sum(weights) == 1`? We can also print a warning if that condition is not fulfilled.

File: docs/source/developer_guides/model_merging.md [hunk 6]
```diff
@@ -138,3 +138,14 @@ print(tokenizer.decode(outputs[0]))
 
 </hfoption>
 </hfoptions>
+
+
+## Merging $(IA)^3$ Models
```
Comment (BenjaminBossan) [inline docs/source/developer_guides/model_merging.md:143]:
```suggestion
## Merging (IA)³ Models
```

LaTeX syntax doesn't seem to be supported: https://moon-ci-docs.huggingface.co/docs/peft/pr_1701/en/developer_guides/model_merging#merging-ia3-models

File: docs/source/developer_guides/model_merging.md [hunk 7]
```diff
@@ -138,3 +138,14 @@ print(tokenizer.decode(outputs[0]))
...
+$(IA)^3$ models support linear model merging. To merge $(IA)^3$ models, you can use the `~IA3Model.add_weighted_adapter` method. This method is similar to the `~LoraModel.add_weighted_adapter` method, but it doesn't accept the `combination_type` parameter. Assuming we have a PEFT model and three $(IA)^3$ adapters, we can merge them as follows:
```
Comment (BenjaminBossan) [inline docs/source/developer_guides/model_merging.md:144]:
```suggestion
(IA)³ models support linear model merging. To merge (IA)³ models, you can use the [`~IA3Model.add_weighted_adapter`] method. This method is similar to the [`~LoraModel.add_weighted_adapter`] method, but it doesn't accept the `combination_type` parameter. Assuming we have a PEFT model and three (IA)³ adapters, we can merge them as follows:
```

File: tests/testing_common.py [hunk 8]
```diff
@@ -1338,6 +1327,63 @@ def _test_weighted_combination_of_adapters(self, model_id, config_cls, config_kw
...
+
+        adapter_list = ["adapter1", "adapter_2", "adapter_3"]
+        weight_list = [0.5, 1.5, 1.5]
+        # Initialize the config
+        config = config_cls(
+            base_model_name_or_path=model_id,
+            **config_kwargs,
+        )
+
+        # Define a dictionary to map config types to their respective test functions
```
Comment (BenjaminBossan) [inline tests/testing_common.py:1370]: I think it's more straightforward to read (though less elegant) if we do:

```python
if isinstance(config, LoraConfig):
    self._test_weighted_combination_of_adapters_lora(...)
elif ...
```

WDYT?

File: tests/testing_common.py [hunk 9]
```diff
@@ -1213,7 +1198,11 @@ def _test_weighted_combination_of_adapters(self, model_id, config_cls, config_kw
 
         # test ties re-weighting with multiple adapters
         model.add_weighted_adapter(
-            adapter_list[:2], weight_list[:2], "multi_adapter_ties_reweighting", combination_type="ties", density=0.5
+            adapter_list[:2],
```
Comment (BenjaminBossan) [inline tests/testing_common.py:1201]: This line is 117 chars, so there should be no need to break it (we have 120 limit).

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the updates, I think we're almost good, I only had a few minor points.

> Maybe we should just specify in the docs that we recommend sum(weights) == 1?

Yes, my suggestion is to document that users should take care that the weights should sum up to 1 or at least be close to 1.

File: tests/testing_common.py (hunk 8 above)
Comment (BenjaminBossan) [reply tests/testing_common.py:1366]: I see what you mean but honestly, for a test I don't care so much about such issues, it's more important to me that I can quickly read from top to bottom and understand what's going on.

**Review by BenjaminBossan — APPROVED** [review]
Thanks a lot Alejandro, LGTM, good to have this feature finally.

@pacman100 do you also want to review?

**Review by pacman100 — APPROVED** [review]
Thank you, @alexrs, for adding support for combining adapters when usiung IA3!  ✨
